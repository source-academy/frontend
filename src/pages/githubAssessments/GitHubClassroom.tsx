import { NonIdealState, Spinner } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { Octokit } from '@octokit/rest';
import { GetResponseDataTypeFromEndpointMethod } from '@octokit/types';
import * as React from 'react';
import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { Redirect, Route, Switch, useLocation } from 'react-router-dom';

import { OverallState } from '../../commons/application/ApplicationTypes';
import ContentDisplay from '../../commons/ContentDisplay';
import { MissionRepoData } from '../../commons/githubAssessments/GitHubMissionTypes';
import GitHubAssessmentsNavigationBar from '../../commons/navigationBar/subcomponents/GitHubAssessmentsNavigationBar';
import { showWarningMessage } from '../../commons/utils/NotificationsHelper';
import { assessmentTypeLink } from '../../commons/utils/ParamParseHelper';
import GitHubAssessmentListing from './GitHubAssessmentListing';
import GitHubAssessmentWorkspaceContainer from './GitHubAssessmentWorkspaceContainer';
import GitHubClassroomWelcome from './GitHubClassroomWelcome';

type DispatchProps = {
  handleGitHubLogIn: () => void;
  handleGitHubLogOut: () => void;
};

/**
 * A page that lists the missions available to the authenticated user.
 * This page should only be reachable if using a GitHub-hosted deployment.
 */
const GitHubClassroom: React.FC<DispatchProps> = props => {
  const location = useLocation<{
    courses: string[] | undefined;
    assessmentTypeOverviews: GHAssessmentTypeOverview[] | undefined;
    selectedCourse: string | undefined;
  }>();
  const octokit: Octokit | undefined = useSelector(
    (store: OverallState) => store.session.githubOctokitObject
  ).octokit;
  const [courses, setCourses] = useState<string[] | undefined>(location.state?.courses);
  const [selectedCourse, setSelectedCourse] = useState<string>(
    location.state?.selectedCourse || ''
  );
  const [assessmentTypeOverviews, setAssessmentTypeOverviews] = useState<
    GHAssessmentTypeOverview[] | undefined
  >(location.state?.assessmentTypeOverviews);
  const types = assessmentTypeOverviews?.map(overview => overview.typeName);

  useEffect(() => {
    if (octokit === undefined) {
      return;
    }

    if (!courses) {
      fetchCourses(octokit, setCourses, setSelectedCourse, setAssessmentTypeOverviews);
    }
  }, [courses, octokit]);

  const changeCourseHandler = React.useCallback(
    (e: any) => {
      if (octokit === undefined) {
        return;
      }

      fetchAssessmentOverviews(octokit, e.target.innerText, setAssessmentTypeOverviews);
      setSelectedCourse(e.target.innerText);
    },
    [octokit, setSelectedCourse, setAssessmentTypeOverviews]
  );

  const refreshAssessmentOverviews = () => {
    if (octokit === undefined) {
      return;
    }

    fetchAssessmentOverviews(octokit, selectedCourse, setAssessmentTypeOverviews);
  };

  const redirectToLogin = () => <Redirect to="/githubassessments/login" />;
  const redirectToAssessments = () => (
    <Redirect
      to={{
        // Types should exist whenever we redirect to assessments as this redirect is only called
        // when a course exists. Unless the course.json is configured wrongly.
        pathname: `/githubassessments/${assessmentTypeLink(types ? types[0] : 'welcome')}`,
        state: {
          courses: courses,
          assessmentTypeOverviews: assessmentTypeOverviews,
          selectedCourse: selectedCourse
        }
      }}
    />
  );

  return (
    <div className="Academy" style={{ overflow: 'hidden' }}>
      <GitHubAssessmentsNavigationBar
        changeCourseHandler={changeCourseHandler}
        handleGitHubLogIn={props.handleGitHubLogIn}
        handleGitHubLogOut={() => {
          props.handleGitHubLogOut();
          setCourses(undefined);
          setAssessmentTypeOverviews(undefined);
          setSelectedCourse('');
        }}
        octokit={octokit}
        courses={courses}
        selectedCourse={selectedCourse}
        types={types}
        assessmentTypeOverviews={assessmentTypeOverviews}
      />
      <Switch>
        <Route
          path="/githubassessments/login"
          render={() => {
            return octokit && (!courses || (courses.length > 0 && !assessmentTypeOverviews)) ? (
              <ContentDisplay
                display={<NonIdealState description="Loading..." icon={<Spinner />} />}
                loadContentDispatch={() => {}}
              />
            ) : octokit && courses && courses.length === 0 ? (
              <Redirect to="/githubassessments/welcome" />
            ) : octokit ? (
              redirectToAssessments()
            ) : (
              <ContentDisplay
                display={
                  <NonIdealState
                    description="Please sign in to GitHub."
                    icon={IconNames.WARNING_SIGN}
                  />
                }
                loadContentDispatch={() => {}}
              />
            );
          }}
        />
        <Route
          path="/githubassessments/welcome"
          component={() => (octokit ? <GitHubClassroomWelcome /> : redirectToLogin())}
        />
        <Route path="/githubassessments/editor" component={GitHubAssessmentWorkspaceContainer} />
        {octokit
          ? types?.map((type, idx) => {
              const filteredAssessments = assessmentTypeOverviews
                ? assessmentTypeOverviews[idx].assessments
                : undefined;
              return (
                <Route
                  path={`/githubassessments/${assessmentTypeLink(type)}`}
                  render={() => (
                    <GitHubAssessmentListing
                      assessmentOverviews={filteredAssessments}
                      refreshAssessmentOverviews={refreshAssessmentOverviews}
                    />
                  )}
                  key={idx}
                />
              );
            })
          : null}
        <Route
          render={
            octokit && courses && courses.length > 0 ? redirectToAssessments : redirectToLogin
          }
        />
      </Switch>
    </div>
  );
};

/**
 * Retrieves list of organizations for the authenticated user.
 *
 * @param octokit The Octokit instance for the authenticated user
 * @param setCourses The React setter function for an array of courses string names
 * @param setSelectedCourse The React setter function for string name of selected course
 */
async function fetchCourses(
  octokit: Octokit,
  setCourses: (courses: string[]) => void,
  setSelectedCourse: (course: string) => void,
  setAssessmentTypeOverviews: (assessmentTypeOverviews: GHAssessmentTypeOverview[]) => void
) {
  const courses: string[] = [];
  const results = (await octokit.orgs.listForAuthenticatedUser({ per_page: 100 })).data;
  const orgs = results.filter(org => org.login.includes('source-academy-course')); // filter only organisations with 'source-academy-course' in name
  orgs.forEach(org => {
    courses.push(org.login.replace('source-academy-course-', ''));
  });
  setCourses(courses);
  if (courses.length > 0) {
    setSelectedCourse(courses[0]);
    fetchAssessmentOverviews(octokit, courses[0], setAssessmentTypeOverviews);
  }
}

export type GHAssessmentTypeOverview = {
  typeName: string;
  assessments: GHAssessmentOverview[];
};

export type GHAssessmentOverview = {
  title: string;
  coverImage: string;
  webSummary: string;
  missionRepoData: MissionRepoData;
  dueDate: Date;
  link?: string;
};

type GitHubAssessment = {
  id: string;
  title: string;
  openAt: string;
  closeAt: string;
  published: string;
  coverImage: string;
  shortSummary: string;
  acceptLink: string;
  repoPrefix: string;
};

async function fetchAssessmentOverviews(
  octokit: Octokit,
  selectedCourse: string,
  setAssessmentTypeOverviews: (assessmentTypeOverviews: GHAssessmentTypeOverview[]) => void
) {
  const userLogin = (await octokit.users.getAuthenticated()).data.login;
  const orgLogin = 'source-academy-course-'.concat(selectedCourse);
  type ListForAuthenticatedUserData = GetResponseDataTypeFromEndpointMethod<
    typeof octokit.repos.listForAuthenticatedUser
  >;
  const userRepos: ListForAuthenticatedUserData = (
    await octokit.repos.listForAuthenticatedUser({ per_page: 100 })
  ).data;
  const courseRepos = userRepos.filter(repo => repo.owner!.login === orgLogin);
  const courseInfoRepo = courseRepos.find(repo => repo.name.includes('course-info'));

  if (courseInfoRepo === undefined) {
    showWarningMessage('The course-info repository cannot be located.', 2000);
    return;
  }

  const files = (
    await octokit.repos.getContent({
      owner: courseInfoRepo.owner!.login,
      repo: courseInfoRepo.name,
      path: ''
    })
  ).data;

  if (Array.isArray(files)) {
    if (files.find(file => file.name === 'course-info.json')) {
      const result = await octokit.repos.getContent({
        owner: courseInfoRepo.owner!.login,
        repo: courseInfoRepo.name,
        path: 'course-info.json'
      });

      const courseInfo = JSON.parse(Buffer.from((result.data as any).content, 'base64').toString());

      courseInfo.types.forEach((type: { typeName: string; assessments: [GitHubAssessment] }) => {
        const assessmentOverviews: GHAssessmentOverview[] = [];
        type.assessments.forEach((assessment: GitHubAssessment) => {
          if (!assessment.published) {
            return;
          }

          const prefixLogin = assessment.repoPrefix + '-' + userLogin;
          const missionRepo = userRepos.find(repo => repo.name === prefixLogin);

          let createdAt = new Date();
          let acceptLink = undefined;
          if (missionRepo === undefined) {
            acceptLink = assessment.acceptLink;
          } else {
            if (missionRepo.created_at !== null) {
              createdAt = new Date(missionRepo.created_at);
            }
          }

          assessmentOverviews.push({
            title: assessment.title,
            coverImage: assessment.coverImage,
            webSummary: assessment.shortSummary,
            missionRepoData: {
              repoOwner: courseInfoRepo.owner!.login,
              repoName: prefixLogin,
              dateOfCreation: createdAt
            },
            dueDate: new Date(assessment.closeAt),
            link: acceptLink
          });

          assessmentOverviews.sort((a, b) => (a.dueDate <= b.dueDate ? -1 : 1));
        });
        (type as any).assessments = assessmentOverviews;
      });
      setAssessmentTypeOverviews(courseInfo.types);
    } else {
      showWarningMessage('The course-info.json file cannot be located.', 2000);
      return;
    }
  }
}

export default GitHubClassroom;
