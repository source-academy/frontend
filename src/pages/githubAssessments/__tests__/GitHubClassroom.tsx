import { act } from '@testing-library/react';
import { shallow } from 'enzyme';
import { useSelector } from 'react-redux';

import GitHubClassroom from '../GitHubClassroom';

type objectPerPage = {
  per_page: number;
};

async function listOrgsForAuthenticatedUser(orgProp: objectPerPage) {
  return {
    data: [
      { login: 'source-academy-course-test' },
      { login: 'not' },
      { login: 'source-academy-course-second' },
      { login: 'valid' },
      { login: 'source-academy-course-third' }
    ]
  };
}

async function listReposForAuthenticatedUser(repoProp: objectPerPage) {
  return {
    data: [
      { name: 'course-info', owner: { login: 'source-academy-course-test' } },
      { name: 'sa-test-mission', owner: { login: 'source-academy-course-test' } },
      { name: 'sa-demo-mission', owner: { login: 'source-academy-course-test' } },
      { name: 'sa-autotester-test', owner: { login: 'source-academy-course-test' } }
    ]
  };
}

const mockCourseInfo =
  'ewogICJDb3Vyc2VOYW1lIjogIkNTMTEwMVMiLAogICJ0eXBlcyI6CiAgWwogICAgewogICAgICAidHlwZU5hbWUiOiAiTm90TWlzc2lvbnMiLAogICAgICAiYXNzZXNzbWVudHMiOgogICAgICBbCiAgICAgICAgewogICAgICAgICAgImlkIjogIjEiLAogICAgICAgICAgInRpdGxlIjogIkN1cnZlIEludHJvZHVjdGlvbiIsCiAgICAgICAgICAib3BlbkF0IjogIjIwMjAtMTItMDFUMDA6MDA6MDArMDg6MDAiLAogICAgICAgICAgImNsb3NlQXQiOiAiMjAyMS0xMi0zMVQyMzo1OTo1OSswODowMCIsCiAgICAgICAgICAicHVibGlzaGVkIjogInllcyIsCiAgICAgICAgICAiY292ZXJJbWFnZSI6ICJodHRwczovL2kuaW1ndXIuY29tL3EyTzRpd2EucG5nIiwKICAgICAgICAgICJzaG9ydFN1bW1hcnkiOiAiSW4gdGhpcyBtaXNzaW9uLCB5b3UgZ2V0IGludHJvZHVjZWQgdG8gdmlzaWJsZSBmdW5jdGlvbnMsIGNhbGxlZCBDdXJ2ZXMhIiwKICAgICAgICAgICJhY2NlcHRMaW5rIjogImh0dHBzOi8vY2xhc3Nyb29tLmdpdGh1Yi5jb20vYS9QeUFVaGRmZSIsCiAgICAgICAgICAicmVwb1ByZWZpeCI6ICJzYS10ZXN0LW1pc3Npb24iCiAgICAgICAgfSwKICAgICAgICB7CiAgICAgICAgICAiaWQiOiAiMiIsCiAgICAgICAgICAidGl0bGUiOiAiRGVtbyBNaXNzaW9uIiwKICAgICAgICAgICJvcGVuQXQiOiAiMjAyMC0xMi0wMVQwMDowMDowMCswODowMCIsCiAgICAgICAgICAiY2xvc2VBdCI6ICIyMDIxLTEyLTMxVDIzOjU5OjU5KzA4OjAwIiwKICAgICAgICAgICJwdWJsaXNoZWQiOiAieWVzIiwKICAgICAgICAgICJjb3ZlckltYWdlIjogImh0dHBzOi8vYXZhdGFycy5naXRodWJ1c2VyY29udGVudC5jb20vdS8zNTYyMDcwNT9zPTQwMCZ1PTMyZjcyZmQxZDY1YTBkNjg3N2FkMWQ1ODcwZmZhMzI3ZGRhNzU0ZjEmdj00IiwKICAgICAgICAgICJzaG9ydFN1bW1hcnkiOiAiUXVpY2tzb3J0IGFzc2lnbm1lbnQgZGVzY3JpcHRpb24hIiwKICAgICAgICAgICJhY2NlcHRMaW5rIjogImh0dHBzOi8vY2xhc3Nyb29tLmdpdGh1Yi5jb20vYS9DeGxxakxhUCIsCiAgICAgICAgICAicmVwb1ByZWZpeCI6ICJzYS1kZW1vLW1pc3Npb24iCiAgICAgICAgfSwKICAgICAgICB7CiAgICAgICAgICAiaWQiOiAiMyIsCiAgICAgICAgICAidGl0bGUiOiAiU29ydGluZyBUaGluZ3MgT3V0IiwKICAgICAgICAgICJvcGVuQXQiOiAiMjAyMC0xMi0wMVQwMDowMDowMCswODowMCIsCiAgICAgICAgICAiY2xvc2VBdCI6ICIyMDIxLTEyLTMxVDIzOjU5OjU5KzA4OjAwIiwKICAgICAgICAgICJwdWJsaXNoZWQiOiAieWVzIiwKICAgICAgICAgICJjb3ZlckltYWdlIjogIi8vczMtYXAtc291dGhlYXN0LTEuYW1hem9uYXdzLmNvbS9taXNzaW9uLWFzc2V0cy9taXNzaW9ucy9RdWlja3NvcnQucG5nIiwKICAgICAgICAgICJzaG9ydFN1bW1hcnkiOiAiQSBxdWljayBsb29rIGF0IHF1aWNrc29ydC4iLAogICAgICAgICAgImFjY2VwdExpbmsiOiAiaHR0cHM6Ly9jbGFzc3Jvb20uZ2l0aHViLmNvbS9hL0QxNmhXdmpBIiwKICAgICAgICAgICJyZXBvUHJlZml4IjogInNhLWF1dG90ZXN0ZXItdGVzdCIKICAgICAgICB9CiAgICAgIF0KICAgIH0sCiAgICB7CiAgICAgICJ0eXBlTmFtZSI6ICJOb3RRdWVzdHMiLAogICAgICAiYXNzZXNzbWVudHMiOgogICAgICBbCiAgICAgICAgewogICAgICAgICAgImlkIjogIjQiLAogICAgICAgICAgInRpdGxlIjogIkZha2UgUXVlc3QiLAogICAgICAgICAgIm9wZW5BdCI6ICIyMDIwLTEyLTAxVDAwOjAwOjAwKzA4OjAwIiwKICAgICAgICAgICJjbG9zZUF0IjogIjIwMjEtMTItMzFUMjM6NTk6NTkrMDg6MDAiLAogICAgICAgICAgInB1Ymxpc2hlZCI6ICJ5ZXMiLAogICAgICAgICAgImNvdmVySW1hZ2UiOiAiaHR0cHM6Ly9pLmt5bS1jZG4uY29tL2VudHJpZXMvaWNvbnMvZmFjZWJvb2svMDAwLzAzNy8wMzcvcGFpbnBla29jb3Zlci5qcGciLAogICAgICAgICAgInNob3J0U3VtbWFyeSI6ICJBIGZha2UgcXVlc3QgdGhhdCBzaG91bGQgc2hvdyB1cC4iLAogICAgICAgICAgImFjY2VwdExpbmsiOiAiaHR0cHM6Ly93d3cueW91dHViZS5jb20vd2F0Y2g/dj1NNVZfSVhNZXdsNCIsCiAgICAgICAgICAicmVwb1ByZWZpeCI6ICJzYS10aGlzLWRvZXMtbm90LWV4aXN0IgogICAgICAgIH0KICAgICAgXQogICAgfSwKICAgIHsKICAgICAgInR5cGVOYW1lIjoiTm90UGF0aHMiLAogICAgICAiYXNzZXNzbWVudHMiOltdCiAgICB9LAogICAgewogICAgICAidHlwZU5hbWUiOiJOb3RDb250ZXN0cyIsCiAgICAgICJhc3Nlc3NtZW50cyI6W10KICAgIH0sCiAgICB7CiAgICAgICJ0eXBlTmFtZSI6Ik90aGVycyIsCiAgICAgICJhc3Nlc3NtZW50cyI6W10KICAgIH0KICBdCn0=';

type getContentProp = {
  owner: string;
  repo: string;
  path: string;
};

async function getContent(getContentProp: getContentProp) {
  const owner = getContentProp.owner;
  const repo = getContentProp.repo;
  const path = getContentProp.path;
  if (owner === 'source-academy-course-test' && repo === 'course-info') {
    if (path === '' || path === undefined) {
      return {
        data: [{ name: 'course-info.json' }]
      };
    }
    if (path === 'course-info.json') {
      return {
        data: {
          content: mockCourseInfo
        }
      };
    }
  }
  return {
    data: [{ name: 'not' }, { name: 'important' }]
  };
}

async function getAuthenticated() {
  return {
    data: {
      login: 'Fubuki'
    }
  };
}

const mockStore = {
  session: {
    githubOctokitObject: {
      octokit: {
        orgs: {
          listForAuthenticatedUser: listOrgsForAuthenticatedUser
        },
        repos: {
          listForAuthenticatedUser: listReposForAuthenticatedUser,
          getContent: getContent
        },
        users: {
          getAuthenticated: getAuthenticated
        }
      }
    }
  }
};

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn()
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    state: undefined
  }),
  useParams: () => ({
    selectedType: undefined
  })
}));

describe('GitHubClassroom', () => {
  beforeEach(() => {
    (useSelector as jest.Mock).mockImplementation(callback => {
      return callback(mockStore);
    });
  });

  const mockProps = {
    handleGitHubLogIn: () => {},
    handleGitHubLogOut: () => {}
  };

  it('renders correctly', async () => {
    await act(async () => {
      const tree = shallow(<GitHubClassroom {...mockProps} />);
      expect(tree.debug()).toMatchSnapshot();
    });
  });
});
