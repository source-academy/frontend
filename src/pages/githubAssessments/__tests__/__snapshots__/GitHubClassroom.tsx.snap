// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHubClassroom renders correctly 1`] = `
"<div className=\\"Academy\\" style={{...}}>
  <GitHubAssessmentsNavigationBar changeCourseHandler={[Function (anonymous)]} handleGitHubLogIn={[Function: handleGitHubLogIn]} handleGitHubLogOut={[Function: handleGitHubLogOut]} octokit={{...}} courses={[undefined]} selectedCourse=\\"\\" types={[undefined]} assessmentTypeOverviews={[undefined]} />
  <Switch>
    <Route path=\\"/githubassessments/login\\" render={[Function: render]} />
    <Route path=\\"/githubassessments/welcome\\" component={[Function: component]} />
    <Route path=\\"/githubassessments/editor\\" component={[Function: C] { displayName: 'withRouter(Connect(GitHubAssessmentWorkspace))', WrappedComponent: [Function: GitHubAssessmentWorkspace], propTypes: { wrappedComponentRef: [Function: bound checkType] { isRequired: [Function: bound checkType] } } }} />
    <Route render={[Function: redirectToLogin]} />
  </Switch>
</div>"
`;
